### DISCUSSION TITLE
Why attention masks are not used in the SD3 self-attention computation?

### DISCUSSION THREAD

**Discussion by reminisce** [huggingface/diffusers#8628] [Q&A]
In the SD3 attention implementation below, why attention masks are not used to mask out text padding tokens in the self-attention computation? Asking this because I'm assuming the text prompts in a batch may have different lengths; we would have to pad shorter sequences to the max length, but we want to mask out padding tokens in the self-attention computation so that they don't affect model inference outputs.

https://github.com/huggingface/diffusers/blob/298ce679992e439a9869632579a2f853ea2b095e/src/diffusers/models/attention_processor.py#L1135

**Comment by sayakpaul** [comment]
First of all, the text embeddings don't interfere with the image ones, as seen in cross-attention. They are processed separately as introduced in the MMDiT architecture of SD3. 

So, text embeddings are passed through self-attention layers. The text embeddings come as a concatenated version of sub-embeddings coming from three different text encoders. So, I would assume implementing a masking scheme would probably be not that straightforward. 

The reference implementation also doesn't mask, either:
https://github.com/Stability-AI/sd3-ref

**Comment by reminisce** [comment]
Thanks @sayakpaul.

According to the implementation, the text and image tokens attend each other in attention as they are concatenated on the sequence dimension before being fed to the `scaled_dot_product_attention` (see the code below). The paper of SD3 also confirms this (see the quote below).

In this sense, the padding tokens (if there are any) of text sequences will take part in attention calculation (resulting in non-zero attention scores). This is odd as different values of `max_seq_length` would affect diffusion model outputs for the text prompts with padding tokens. Although implementing a masking scheme is not as simple as in other architectures where only one text encoder is used, I think it is still feasible to do it for the MMDiT in SD3 (the mask will have interleaved `0`s and `-inf`s).

https://github.com/huggingface/diffusers/blob/e7b9a0762bd015372ea56767809bc2132f6b4dad/src/diffusers/models/attention_processor.py#L1020-L1030

> Quote from SD3 paper Section 4: this is equivalent to
having two independent transformers for each modality, but
joining the sequences of the two modalities for the **attention
operation**, such that both representations can work in their
own space yet **take the other one into account**.

**Reply by sayakpaul** [reply]
Oh you are 100% right. If you want to take a stab at implementing it, feel free to -- we will be happy to help.

**Reply by weathon** [reply]
I have the same question, I am trying to modify the attention and I saw that there is no masks, so I am very confused

**Comment by reminisce** [comment]
Thanks for confirming this.  It would be good to create an issue to get help from the community and track the progress (sorry that I may not have bandwidth to implement the fix).

**Reply by sayakpaul** [reply]
Sure feel free to create one.

**Reply by reminisce** [reply]
Issue created: https://github.com/huggingface/diffusers/issues/8673.

Thanks for the discussion and feel free to close this thread.
